### PR TITLE
Faster room joins: Resume state re-syncing after a Synapse restart

### LINKED ISSUE #12642: Faster joins: resume state re-syncing after a Synapse restart
Currently, if Synapse is restarted in the middle of a state re-sync, it won't resume the work.

### PR DESCRIPTION
**Pull request by squahtx**
Resolves #12642.
Depends on #12812.

~~Tested manually. I'm stuck on how to test this automatically.~~
Tested by https://github.com/matrix-org/complement/pull/378

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -1466,6 +1485,7 @@ async def _sync_partial_state_room(
                 `destination` is unavailable
             room_id: room to be resynced
         """
+        assert not self.config.worker.worker_app
```
Comment (squahtx) [inline synapse/handlers/federation.py:1488]: This is implied by the identical assert in `do_invite_join`, which starts the background task.

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -1452,6 +1457,20 @@ async def get_room_complexity(
...
+
+        partial_state_rooms = await self.store.get_partial_state_rooms_and_servers()
+        for room_id, servers_in_room in partial_state_rooms.items():
+            run_as_background_process(
+                desc="sync_partial_state_room",
+                func=self._sync_partial_state_room,
+                destination=None,
+                destinations=servers_in_room,
+                room_id=room_id,
+            )
```
Comment (squahtx) [inline synapse/handlers/federation.py:1472]: I could not see a simple way to determine the `destination` the original sync was using after a restart, so I took the entire list of servers in the room to try.

File: synapse/handlers/federation.py [hunk 3]
```diff
@@ -168,6 +168,11 @@ def __init__(self, hs: "HomeServer"):
 
         self.third_party_event_rules = hs.get_third_party_event_rules()
 
+        if not hs.config.worker.worker_app:
+            run_as_background_process(
+                "resume_sync_partial_state_room", self._resume_sync_partial_state_room
+            )
```
Comment (richvdh) [inline synapse/handlers/federation.py:174]: Could do with a comment here:

```suggestion
        # if this is the main process, fire off a background process to resume
        # any partial-state-resync operations which were in flight when we
        # were shut down.
        if not hs.config.worker.worker_app:
            run_as_background_process(
                "resume_sync_partial_state_room", self._resume_sync_partial_state_room
            )
```

File: synapse/storage/databases/main/room.py [hunk 4]
```diff
@@ -1077,6 +1078,34 @@ def get_rooms_for_retention_period_in_range_txn(
...
+            retcols=("room_id", "server_name"),
+            desc="get_partial_state_rooms",
+        )
+
+        for row in rows:
+            room_id = row["room_id"]
+            server_name = row["server_name"]
+            if room_id not in room_servers:
+                room_servers[room_id] = []
+            room_servers[room_id].append(server_name)
```
Comment (richvdh) [inline synapse/storage/databases/main/room.py:1105]: this is normally spelt:

```suggestion
            room_servers.setdefault(room_id, []).append(server_name)
```

File: synapse/storage/databases/main/room.py [hunk 5]
```diff
@@ -1077,6 +1078,34 @@ def get_rooms_for_retention_period_in_range_txn(
...
+
+        Returns:
+            A dictionary of rooms with partial state, with room IDs as keys and
+            lists of servers in rooms as values.
+        """
+        room_servers: Dict[str, List[str]] = {}
+
+        rows = await self.db_pool.simple_select_list(
+            "partial_state_rooms_servers",
+            keyvalues={},
```
Comment (richvdh) [inline synapse/storage/databases/main/room.py:1095]: might as well save creating the empty dict here:

```suggestion
            keyvalues=None,
```

**Review by richvdh — APPROVED** [review]
lgtm, a few hairs to split

File: synapse/handlers/federation.py (hunk 2 above)
Comment (richvdh) [reply synapse/handlers/federation.py:1472]: yeah, I think this is ok for now, but it definitely leads to less than satisfactory results while it works its way through the entire server list of Matrix HQ trying to find a server which will respond.

**Pull request merged**